README.md
# Implementing a Monte Carlo Simulation in Haskell User Guide
Hello this is a guide on how to successfully run our project's program please read the entire document, skipping sections only when you already know that they apply to you. This installation process was verified on a blank Windows 10 machine that had nothing related installed on it. 

# Install Haskell (with Stack and Cabal)
Install the Haskell programming language via ghccup with this link: https://www.haskell.org/ghcup/ follow whichever install guide is appropriate for your system, the importpant part is to be sure you install the following: Haskell, stack, cabal.

# Install Git
Our project is hosted on GitHub, so it uses Git as the version control system, you need to make sure you have this installed as well download is available at the following link: https://git-scm.com/downloads

# Clone The Repository
If you are already reading this on the GitHub repository main page then click the green code button towards the top and copy the http link provided, open a terminal and navigate to where you'd like the project to be cloned to and the run the following command: <code>git clone x</code> with <code>x</code> being the link you copied.

If you are reading this from the README present in the file structure of this project from an alternative source, visit the GitHub at the following link and follow the above instruction: https://github.com/griffinnewbold/COMS-4995-Project

# Navigate into the monteParSim directory
Once you have cloned the repository run the following command: <code>cd COMS-4995-Project\monteParSim</code>

# Build using <code>stack build</code>
If it is the very first time doing this then it will take a bit of time, you shouldn't have any issues but in the event you do visit the <code>stack.yaml</code> file and scroll down to <code>extra-deps</code> and install the following packages manually with the <code>stack install package</code> command the package does not include any numbers just characters belonging to the english alphabet.

You can execute the main method of the program and specify relative alternative options with the following: <code>stack build --exec "monteParSim-exe +RTS -N4"</code> this command enables runtime system options and runs with 4 cores at the program's disposal, this is only relevant for the parallel portions of the code.  

# Verify Functionality using <code>stack test</code>
To ensure nothing broke prior to executing the code on your own, you can run this command to verify via our unit tests that the code we wrote 
1. Does not error out 
2. Provides accurate values 

Just like with the build command you can specify additional options with the following: <code>stack test --test-arguments "+RTS -N4"</code> this command enables runtime system options and runs with 4 cores at the program's disposal, this is only relevant for the parallel portions of the code.  


monteParSim/CHANGELOG.md
We are required to have this, move on.

Test.hs
module Test (bernoulli, monteCarloAsianParallel, unfoldsSMGen) where

import System.Random.SplitMix
import Control.Parallel.Strategies

-- A function to generate a Bernoulli trial result given a probability and a generator
bernoulli :: Double -> SMGen -> (Int, SMGen)
bernoulli p gen = let (random_val, gen') = nextDouble gen
                  in (if random_val < p then 1 else 0, gen')

-- The main Monte Carlo function
monteCarloAsianParallel :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> Double
monteCarloAsianParallel n t r u d s0 k =
  let discount = 1 / ((1 + r) ^ t)
      p_star = (1 + r - d) / (u - d)
      initialGen = mkSMGen 42 -- Seed for the random number generator

      -- Function to calculate the trial value
      calcPrice :: Int -> Double -> Double -> SMGen -> (Double, SMGen)
      calcPrice i sum_prices price genCalc
        | i == t    = (sum_prices, genCalc)
        | otherwise = let (b, genNext) = bernoulli p_star genCalc
                          (newPrice, newGen) = if b == 1
                                               then (price * u, genNext)
                                               else (price * d, genNext)
                      in calcPrice (i + 1) (sum_prices + newPrice) newPrice newGen

      -- Function for a single trial
      trial :: SMGen -> Double
      trial genTrial = let (sum_prices, _) = calcPrice 0 0 s0 genTrial
                           diff_val = (sum_prices / fromIntegral t) - k
                       in max diff_val 0

  in (sum (parMap rdeepseq trial (unfoldsSMGen initialGen n)) * discount) / fromIntegral n

-- Helper function to unfold SMGen into a list of n generators
unfoldsSMGen :: SMGen -> Int -> [SMGen]
unfoldsSMGen gen n = take n $ iterate (snd . splitSMGen) gen


monteParSim/test/Spec.hs
import Library 
import Test.HUnit
import System.Random.SplitMix
import Control.Concurrent (getNumCapabilities)


{- | 
Test Suite for the Monte Carlo Simulations functions contained within the Library module.

Precondition: HUnit library is installed, use stack install hunit to confirm installation or proceed with it. 
Postcondition: All Tests run successfully.
-}

test1 :: Test
test1 = TestCase $ do
  let n = 10000
      t = 10
      r = 0.05
      u = 1.15
      d = 1.01
      s0 = 50
      k = 70

  result <- monteCarloAsian n t r u d s0 k

  let lowerBound = 0.9 
      upperBound = 1.4  

  assertBool "Result is within bounds" (result >= lowerBound && result <= upperBound)

test2 :: Test
test2 = TestCase $ do
  let n = 10000
      t = 10
      r = 0.05
      u = 1.15
      d = 1.01
      s0 = 50
      k = 70

  result <- monteCarloAsianVector n t r u d s0 k

  let lowerBound = 0.9 
      upperBound = 1.4  

  assertBool "Result is within bounds" (result >= lowerBound && result <= upperBound)

test3 :: Test
test3 = TestCase $ do
  let n = 10000
      t = 10
      r = 0.05
      u = 1.15
      d = 1.01
      s0 = 50
      k = 70

  numCores <- getNumCapabilities
  initGen <- initSMGen
  let result = monteCarloAsianParallel numCores n t r u d s0 k initGen

  let lowerBound = 0.9 
      upperBound = 1.4  

  assertBool "Result is within bounds" (result >= lowerBound && result <= upperBound)

test4 :: Test
test4 = TestCase $ do
  let n = 10000
      t = 10
      r = 0.05
      u = 1.15
      d = 1.01
      s0 = 50
      k = 70

  numCores' <- getNumCapabilities
  initGen' <- initSMGen
  let result = monteCarloAsianParallelVector numCores' n t r u d s0 k initGen'

  let lowerBound = 0.9 
      upperBound = 1.4  

  assertBool "Result is within bounds" (result >= lowerBound && result <= upperBound)


tests :: Test
tests = TestList [TestLabel "Sequential Test [Non Vector]" test1, TestLabel "Sequential Test [Vector]" test2, TestLabel "Parallel Test [Non Vector]" test3, TestLabel "Parallel Test [Vector]" test4]

main :: IO ()
main = do
  results <- runTestTT tests
  print results

monteParSim/performance_test/Test.hs
{-# LANGUAGE BangPatterns #-}
module Test (main, monteCarloAsianParallel) where


import System.Random.SplitMix
import Control.Concurrent (getNumCapabilities)
import Control.Parallel.Strategies
import System.Random
import Control.Monad (replicateM, when, unless)

{-
This is the file allows for testing the individual performance of the
sequential and parallel versions of the Monte Carlo Asian Option Functions
with different parameters/ number of cores.  This file was used to conduct
most of the performance tests seen in the project report.

To use, follow the instructions below:

1. Modify the main by commenting out all but one of the function calls
(monteCarloAsian/monteCarloAsianParallel/monteCarloAsianParallelOld).
The uncommented function call will be ran.  Feel free to modify the function
parameters to test performance on the different parameters

2. While in the performance_test directory, run the command to build the file

 ghc -main-is Test -O2 -rtsopts -eventlog -threaded Test
 
 If the code fails to compile, ensure you have the correct dependencies installed.

3. Run the following to run the executable and generate the eventlog.

./Test +RTS -N1 -l

You can change the flag -N(1/2/3/etc.) to try different core counts.

4. The following command can be used to view the ThreadScope eventlog

threadscope Test.eventlog

-}
main:: IO ()
main = do
    init_gen <- initSMGen
    res <- monteCarloAsian 5000000 100 0.05 1.15 1.01 50 70
--    let res = monteCarloAsianParallel 3 10 100 0.05 1.15 1.01 50 70 init_gen
--    let res = monteCarloAsianParallelOld 5000000 100 0.05 1.15 1.01 50 70 init_gen

    putStrLn $ show res

-- Sequential Content Begins --
monteCarloAsian :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO Double
monteCarloAsian n t r u d s0 k = do
  let discount = 1 / ((1 + r) ^ t)
      pStar    = (1 + r - d) / (u - d)

  let seqTrial = do
        let seqCalcPrice i sumPrices price
              | i == t = return sumPrices
              | otherwise = do
                b <- bernoulli pStar
                if b == 1
                    then seqCalcPrice (i + 1) (sumPrices + (price * u)) (price * u)
                    else seqCalcPrice (i + 1) (sumPrices + (price * d)) (price * d)
        sumPrices <- seqCalcPrice 0 0.0 s0
        let diffVal = (sumPrices / fromIntegral t) - k
        return $ diffVal `max` 0

  total <- sum <$> replicateM n seqTrial
  return $ (total * discount) / fromIntegral n

-- Helper function for sequential algorithm to generate random values
bernoulli :: Double -> IO Int
bernoulli p = do
    randomVal <- randomIO -- :: IO Double
    return $ if randomVal < p then 1 else 0

-- Sequential Content Ends --

-- Parallel Content Begins --
  
-- Main function for the parallel simulations
monteCarloAsianParallel :: Int -> Int -> Int -> Double -> Double -> Double -> Double -> Double -> SMGen -> Double
monteCarloAsianParallel numCores n t r u d s0 k init_gen =
  let !discount = 1 / ((1 + r) ^ t)
      !p_star = (1 + r - d) / (u - d)
      chunkSize = n `div` (10 * numCores)
      gens = unfoldsSMGen init_gen n
      trials = withStrategy (parListChunk chunkSize rdeepseq) $
               map (runEval . trial p_star u d s0 k t) gens
      !result = sum trials * discount / fromIntegral n
  in result

-- Helper function to unfold SMGen into a list of n generators
unfoldsSMGen :: SMGen -> Int -> [SMGen]
unfoldsSMGen gen n = take n $ iterate (snd . splitSMGen) gen

-- Helper function to complete a single trial
trial :: Double -> Double -> Double -> Double -> Double -> Int -> SMGen -> Eval Double
trial p_star u d s0 k t genTrial = do
  let (!sum_prices, _) = calcPrice 0 t p_star u d 0 s0 genTrial
      !diff_val = (sum_prices / fromIntegral t) - k
  return $ max diff_val 0

-- Helper function to calculate the price in a given trial
calcPrice :: Int -> Int -> Double -> Double -> Double -> Double -> Double ->
            SMGen -> (Double, SMGen)
calcPrice i t p_star u d !sum_prices !price genCalc
  | i == t    = (sum_prices, genCalc)
  | otherwise = let (b, genNext) = bernoulliParallel p_star genCalc
                    (!newPrice, newGen) = if b == 1
                                          then (price * u, genNext)
                                          else (price * d, genNext)
                in calcPrice (i + 1) t p_star u d (sum_prices + newPrice) newPrice newGen

-- Helper function to generate a Bernoulli trial result given a probability and a generator
bernoulliParallel :: Double -> SMGen -> (Int, SMGen)
bernoulliParallel p gen = let (!random_val, gen') = nextDouble gen
                          in (if random_val < p then 1 else 0, gen')

-- Parallel Content Ends --

-- Old Parallel Content --

trialOld :: Double -> Double -> Double -> Double -> Double -> Int -> SMGen -> Double
trialOld p_star u d s0 k t genTrial =
  let (sum_prices, _) = calcPrice 0 t p_star u d 0 s0 genTrial
      diff_val = (sum_prices / fromIntegral t) - k
  in max diff_val 0

monteCarloAsianParallelOld :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> SMGen -> Double
monteCarloAsianParallelOld n t r u d s0 k gen =
  let discount = 1 / ((1 + r) ^ t)
      p_star = (1 + r - d) / (u - d)
      chunkSize = n `div` 4
      gens = unfoldsSMGen gen n
      trials = parMap rdeepseq (trialOld p_star u d s0 k t)
                      gens `using` parListChunk chunkSize rdeepseq
      result = sum trials * discount / fromIntegral n
  in result


monteParSim/src/Lib.hs
-- module Lib (someFunc, randomFloat, bernoulli, exactPrice, seqMonte) where
module Lib (exactPrice) where

-- import System.Random
import Control.Monad (replicateM)
import Data.Functor
-- import Math.Combinatorics.Exact

-- bernoulli :: Float -> IO Int
-- bernoulli p = do
--     random_val <- randomFloat
--     return $ if random_val < p then 1 else 0

-- randomFloat :: IO Float
-- randomFloat = do
--     gen <- newStdGen
--     let (randomNumber, _) = randomR (0.0, 1.0) gen :: (Float, StdGen)
--     return randomNumber

-- seqMonte 0 0 r u d s_o k = (1::Float)
-- seqMonte 0 t r u d s_o k = (1::Float)
-- seqMonte n t r u d s_o k = (1::Float)


-- --  I am currently working on exact price
-- exact_price t r u d s_o k = p_star
--         where p_star = ((1::Float)+r-d)/(u-d)


binomial :: Integer -> Integer -> Integer
binomial n k
    | k < 0 || k > n = 0
    | otherwise = product [1..n] `div` (product [1..k] * product [1..(n - k)])


{-
PYTHON CODE
def exact_price(T,r,u,d,S0,K):
    total = 0
    p_star= (1+r-d)/(u-d)
    for k in range(T+1):
        S = S0*u**k*d**(T-k)
        total += n_choose_k(T, k)*p_star**k*(1-p_star)**(T-k)*max(S-K, 0)
    total *= (1+r)**-T
    return total

exactPrice 10 0.05 1.15 1.01 50 70
-}

exactPrice :: Integer -> Double -> Double -> Double -> Double -> Double -> Double
exactPrice t r u d s0 k = total * ((1 + r)** negative_t)
  where
    negative_t = (-(fromIntegral t::Double))
    pStar = (1 + r - d) / (u - d)
    total = sum [(fromIntegral (binomial t i)) * (pStar^^i) * ((1 - pStar) ** (fromIntegral(t - i))) * max (s0 * u^i * d**(fromIntegral(t - i)) - k) 0 | i <- [0..t] ]
    -- total = sum [ fromIntegral (binomial t k)) * pStar^^k * (1 - pStar) ** ((fromIntegral(t - k)) * max (s0 * u^k * d**(fromIntegral(t - k))) 0) | k <- [0..t] ]

monteParSim/app/Main.hs
module Main (main) where

import Library
import Control.Concurrent (getNumCapabilities)
import System.Random.SplitMix
import Data.Time

{- |
Entry point for the system asks the user to enter different quantities
and then validates the input prior to execution. 
-}
main :: IO ()
main = do
    putStrLn "Enter the number of simulations (n):"
    n <- read <$> getLine

    putStrLn "Enter the number of time steps (t):"
    t <- read <$> getLine

    putStrLn "Enter the interest rate (r):"
    r <- read <$> getLine

    putStrLn "Enter the up factor (u):"
    u <- read <$> getLine

    putStrLn "Enter the down factor (d):"
    d <- read <$> getLine

    putStrLn "Enter the initial stock price (s0):"
    s0 <- read <$> getLine

    putStrLn "Enter the strike price (k):"
    k <- read <$> getLine

    validateInputs n t r u d s0 k

    -- sequential non vector
    putStrLn "Sequential Monte Carlo Simulation:"
    start <- getCurrentTime
    resultAsian <- monteCarloAsian n t r u d s0 k
    stop <- getCurrentTime
    let timeDiff = diffUTCTime stop start
    putStrLn $ "Result Monte Carlo Asian Option [Sequential]: " ++ show resultAsian
    putStrLn $ "Time to Run: " ++ show timeDiff

    -- parallel non vector
    coreCount <- getNumCapabilities
    initGen <- initSMGen
    start' <- getCurrentTime
    let resultPar = monteCarloAsianParallel coreCount n t r u d s0 k initGen
    stop' <- getCurrentTime
    let timeDiff' = diffUTCTime stop' start'
    putStrLn $ "Result Monte Carlo Asian Option [Parallel]: " ++ show resultPar
    putStrLn $ "Time to Run: " ++ show timeDiff'

    -- sequential vector
    putStrLn "Sequential Monte Carlo Simulation Vector:"
    start'' <- getCurrentTime
    resultAsianVec <- monteCarloAsianVector n t r u d s0 k
    stop'' <- getCurrentTime
    let timeDiff'' = diffUTCTime stop'' start''
    putStrLn $ "Result Monte Carlo Asian Option [Sequential Vector]: " ++ show resultAsianVec
    putStrLn $ "Time to Run: " ++ show timeDiff''

    -- parallel vector
    putStrLn "Monte Carlo Simulation Parallel Vector:"
    start''' <- getCurrentTime
    let resultAsianPA = monteCarloAsianParallelVector coreCount n t r u d s0 k initGen
    stop''' <- getCurrentTime
    let timeDiff''' = diffUTCTime stop''' start'''
    putStrLn $ "Result Monte Carlo Asian Option [Parallel Vector]: " ++ show resultAsianPA
    putStrLn $ "Time to Run: " ++ show timeDiff'''

monteParSim/src/Library.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Library (bernoulli, monteCarloAsian, validateInputs, monteCarloAsianParallel,
bernoulliParallel, unfoldsSMGen, monteCarloAsianVector, monteCarloAsianParallelVector,
unfoldBernoulli, vectorTrial) where

import System.Random
import System.Random.SplitMix
import qualified Data.Vector as V
import Control.Monad (replicateM, unless, when)
import Control.Parallel.Strategies (rdeepseq, Eval, parListChunk, runEval, withStrategy)

-- Sequential Content Begins --
monteCarloAsian :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO Double
monteCarloAsian n t r u d s0 k = do
  -- Calculate discount factor and probability of an up movement
  let discount = 1 / ((1 + r) ^ t)
      pStar    = (1 + r - d) / (u - d)

  -- Perform a single trial of the simulation
  let seqTrial = do
        -- Recursively calculate the sum of prices along a path
        let seqCalcPrice i sumPrices price
              | i == t = return sumPrices
              | otherwise = do
                b <- bernoulli pStar
                if b == 1
                    then seqCalcPrice (i + 1) (sumPrices + (price * u)) (price * u)
                    else seqCalcPrice (i + 1) (sumPrices + (price * d)) (price * d)

        -- Calculate the difference between average simulate price and strike price
        sumPrices <- seqCalcPrice 0 0.0 s0
        let diffVal = (sumPrices / fromIntegral t) - k
        return $ diffVal `max` 0

  -- Perform 'n' trials and compute the average
  total <- sum <$> replicateM n seqTrial
  return $ (total * discount) / fromIntegral n

-- Main function for sequential vectors
monteCarloAsianVector :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO Double
monteCarloAsianVector n t r u d s0 k = do
  let discount = 1 / ((1 + r) ** fromIntegral t)
      pStar    = (1 + r - d) / (u - d)

  let vectorTrialSeq = do
        steps <- V.replicateM t (bernoulli pStar)
        let priceVector = V.scanl' (\price step -> price * (if step == 1 then u else d)) s0 steps
            sumPrices   = V.sum priceVector - V.head priceVector
            avgPrice    = sumPrices / fromIntegral t
            diffVal     = avgPrice - k
        return $ max diffVal 0

  total <- sum <$> replicateM n vectorTrialSeq
  return $ (total * discount) / fromIntegral n

-- Helper function for sequential algorithm to generate random values
bernoulli :: Double -> IO Int
bernoulli p = do
    randomVal <- randomIO -- :: IO Double
    return $ if randomVal < p then 1 else 0

-- Sequential Content Ends --
-- Parallel Content Begins --

-- Main function for the parallel simulations
monteCarloAsianParallel :: Int -> Int -> Int -> Double -> Double -> Double -> Double -> Double -> SMGen -> Double
monteCarloAsianParallel numCores n t r u d s0 k init_gen =
  let !discount = 1 / ((1 + r) ^ t)
      !pStar = (1 + r - d) / (u - d)
      chunkSize = n `div` (10 * numCores)
      gens = unfoldsSMGen init_gen n
      trials = withStrategy (parListChunk chunkSize rdeepseq) $
               map (runEval . trial pStar u d s0 k t) gens
      !result = sum trials * discount / fromIntegral n
  in result

-- Helper function to unfold SMGen into a list of n generators
unfoldsSMGen :: SMGen -> Int -> [SMGen]
unfoldsSMGen gen n = take n $ iterate (snd . splitSMGen) gen

-- Helper function to complete a single trial
trial :: Double -> Double -> Double -> Double -> Double -> Int -> SMGen -> Eval Double
trial pStar u d s0 k t genTrial = do
  let (!sumPrices, _) = calcPrice 0 t pStar u d 0 s0 genTrial
      !diffVal = (sumPrices / fromIntegral t) - k
  return $ max diffVal 0

-- Helper function to calculate the price in a given trial
calcPrice :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> SMGen -> (Double, SMGen)
calcPrice i t pStar u d !sumPrices !price genCalc
  | i == t    = (sumPrices, genCalc)
  | otherwise = let (b, genNext) = bernoulliParallel pStar genCalc
                    (!newPrice, newGen) = if b == 1
                                          then (price * u, genNext)
                                          else (price * d, genNext)
                in calcPrice (i + 1) t pStar u d (sumPrices + newPrice) newPrice newGen

-- Helper function to generate a Bernoulli trial result given a probability and a generator
bernoulliParallel :: Double -> SMGen -> (Int, SMGen)
bernoulliParallel p gen = let (!randomVal, gen') = nextDouble gen
                          in (if randomVal < p then 1 else 0, gen')

-- Main function for vectors in parallel
monteCarloAsianParallelVector :: Int -> Int -> Int -> Double -> Double -> Double -> Double -> Double -> SMGen -> Double
monteCarloAsianParallelVector numCores n t r u d s0 k init_gen =
  let !discount = 1 / ((1 + r) ** fromIntegral t)
      !pStar = (1 + r - d) / (u - d)
      chunkSize = n `div` (10 * numCores)
      gens = unfoldsSMGen init_gen n
      trials = withStrategy (parListChunk chunkSize rdeepseq) $
               map (runEval . vectorTrial pStar u d s0 k t) gens
      !result = sum trials * discount / fromIntegral n
  in result

-- Helper function to complete a single vector trial
vectorTrial :: Double -> Double -> Double -> Double -> Double -> Int -> SMGen -> Eval Double
vectorTrial pStar u d s0 k t gen = do
  let steps = V.unfoldrN t (unfoldBernoulli pStar) gen
      priceVector = V.scanl' (\price step -> price * (if step == 1 then u else d)) s0 steps
      sumPrices = V.sum priceVector - V.head priceVector
      avgPrice = sumPrices / fromIntegral t
      diffVal = avgPrice - k
  return $ max diffVal 0

-- Helper function to generate the bernoullis
unfoldBernoulli :: Double -> SMGen -> Maybe (Int, SMGen)
unfoldBernoulli p gen = Just $ bernoulliParallel p gen
-- Parallel Content Ends --

-- General All purpose helper functions below -- 
-- Helper function to validate the provided inputs from the user
validateInputs :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO ()
validateInputs n t r u d s0 k = do
  when (n <= 0) $ error "Invalid value for n. Number of trials (n) must be greater than 0."
  when (t < 1) $ error "Invalid value for t. Number of time steps (t) must be greater than or equal to 1."
  when (r <= 0) $ error "Invalid value for r. The interest rate (r) must be greater than 0."
  when (u <= 0) $ error "Invalid value for u. The up factor (u) must be greater than 0."
  when (d <= 0) $ error "Invalid value for d. The down factor (d) must be greater than 0."
  when (s0 <= 0) $ error "Invalid value for s0. Initial stock price (s0) must be greater than 0."
  when (k <= 0) $ error "Invalid value for k. Strike price (k) must be greater than 0."
  unless (0 < d && d < 1 + r && 1 + r < u) $
    error "Invalid values for r, u, and d entered.\nThe relationship 0 < d < r < u must be maintained to get valid results."

monteParSim/src/library.hs
 {-# LANGUAGE BangPatterns #-}
{-
stack --resolver lts-19.23 ghci 
stack ghci --package random
:set -Wall
 -}
module Library (bernoulli, monteCarloAsian, validateInputs, rand_lst, monteCarloAsianParallel) where

import System.Random
import Control.Monad (replicateM, unless, when, forM)
import Control.Concurrent (forkIO)
import Control.Parallel.Strategies (parList, rdeepseq, using, rseq)
import System.Random.Mersenne.Pure64


-- I think I need to use state/ something similar so that I can have un updated state for each
-- next generator.  Right now it can just generate a list of random doubles between 0 1.

get_rand_lst :: Int -> [Double] -> PureMT -> [Double]
get_rand_lst 0 !x _ = x
get_rand_lst n !x gen =
    get_rand_lst (n - 1) (y:x) gen'
  where
    (y, gen') = randomDouble gen

rand_lst :: Int -> [Double]
rand_lst n = get_rand_lst n [] (pureMT $ fromIntegral seed)

seed :: Int
seed = 10

bernoulli :: Double -> IO Int
bernoulli p = do
    random_val <- randomIO :: IO Double
    return $ if random_val < p then 1 else 0

{-
Command:
monteCarloAsian 10000 10 0.05 1.15 1.01 50 70
-}
monteCarloAsian :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO Double
monteCarloAsian n t r u d s0 k = do
  validateInputs n t r u d s0 k
  let discount = 1 / ((1 + r) ^ t)
      p_star = (1 + r - d) / (u - d)

  let trial = do
        let calcPrice i sum_prices price
              | i == t = return sum_prices
              | otherwise = do
                b <- bernoulli p_star
                if b == 1
                    then calcPrice (i + 1) (sum_prices + (price*u)) (price*u)
                    else calcPrice (i + 1) (sum_prices + (price*d)) (price*d)
        sum_prices <- calcPrice 0 (0::Double) s0
        {- diff_val: difference between the average stock price and the strike price.-}
        let diff_val = (sum_prices / fromIntegral t :: Double) - k
        return $ max diff_val 0

  total <- sum <$> replicateM n trial
  return $ (total * discount) / fromIntegral n

monteCarloAsianParallel :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO Double
monteCarloAsianParallel n t r u d s0 k = do
  validateInputs n t r u d s0 k
  let discount = 1 / ((1 + r) ^ t)
      p_star = (1 + r - d) / (u - d)

  let trial = do
        let calcPrice i sum_prices price
              | i == t = return sum_prices
              | otherwise = do
                b <- bernoulli p_star
                if b == 1
                    then calcPrice (i + 1) (sum_prices + (price*u)) (price*u)
                    else calcPrice (i + 1) (sum_prices + (price*d)) (price*d)
        sum_prices <- calcPrice 0 0 s0
        let avg_price = sum_prices / fromIntegral t :: Double
        let diff_val = avg_price - k
        return $ max diff_val 0
  total <- sum <$> replicateM n trial
  return $ (total * discount) / fromIntegral n


validateInputs :: Int -> Int -> Double -> Double -> Double -> Double -> Double -> IO ()
validateInputs n t r u d s0 k = do
  when (n <= 0) $ error "Invalid value for n. Number of trials (n) must be greater than 0."
  when (t < 1) $ error "Invalid value for t. Number of time steps (t) must be greater than or equal to 1."
  when (r <= 0) $ error "Invalid value for r. The interest rate (r) must be greater than 0."
  when (u <= 0) $ error "Invalid value for u. The up factor (u) must be greater than 0."
  when (d <= 0) $ error "Invalid value for d. The down factor (d) must be greater than 0."
  when (s0 <= 0) $ error "Invalid value for s0. Initial stock price (s0) must be greater than 0."
  when (k <= 0) $ error "Invalid value for k. Strike price (k) must be greater than 0."
  unless (0 < d && d < 1 + r && 1 + r < u) $
    error "Invalid values for r, u, and d entered.\nThe relationship 0 < d < r < u must be maintained to get valid results."
  
  -- Evaluate trials in parallel
--   total <- sum . parList rseq <$> replicateM n trial
 -- total <- parList rseq <$> replicateM n trial
 -- return $ (sum total * discount) / fromIntegral n
                           

{-
 Thread state 
 Write your own random nums using thread monad

 Threading the states?
 parallel random generator haskell
-}
